innative/atomic_instructions.cpp

```diff
@@ -0,0 +1,256 @@
+    auto varsize = unsigned(4 * -varTy); // TE_i32(-1) => 4; TE_i64(-2) => 8;
+    auto varty   = builder.getIntNTy(varsize * 8);
+
+    auto load = builder.CreateAlignedLoad(ptr, align, name);
+    load->setAtomic(SeqCst);
+
+    llvmVal* result = load;
+    // All atomic loads zero-ext when the atomic variable is smaller than the WASM value
+    if(varsize > align)
+      result = builder.CreateZExt(result, varty);
+
+    return PushReturn(result);
+  });
+}
+
+IN_ERROR innative::Compiler::CompileAtomicStore(Instruction& ins, WASM_TYPE_ENCODING varTy, const char* name)
+{
+  return CompileAtomicMemInstruction(ins, varTy, [&](unsigned align, llvmVal* ptr, llvmVal* value) {
+    auto varsize = unsigned(4 * -varTy); // TE_i32(-1) => 4; TE_i64(-2) => 8;
+    auto memty   = builder.getIntNTy(align * 8);
+
+    // We need to truncate first when the atomic field is smaller than the WASM value
+    if(varsize > align)
+      value = builder.CreateTrunc(value, memty);
+
+    auto store = builder.CreateAlignedStore(value, ptr, align);
+    store->setAtomic(SeqCst);
+
+    return ERR_SUCCESS;
+  });
+}
+
+IN_ERROR innative::Compiler::CompileAtomicRMW(Instruction& ins, WASM_TYPE_ENCODING varTy, llvm::AtomicRMWInst::BinOp Op,
+                                              const char* name)
+{
+  return CompileAtomicMemInstruction(ins, varTy, [&](unsigned align, llvmVal* ptr, llvmVal* newValue) {
+    auto varsize = unsigned(4 * -varTy); // TE_i32(-1) => 4; TE_i64(-2) => 8;
+    auto varty   = builder.getIntNTy(varsize * 8);
+    auto memty   = builder.getIntNTy(align * 8);
+
+    if(varsize > align)
+      newValue = builder.CreateTrunc(newValue, memty);
+
+    auto rmw = builder.CreateAtomicRMW(Op, ptr, newValue, SeqCst);
+    rmw->setName(name);
+
+    llvmVal* oldValue = rmw;
+    if(varsize > align)
+      oldValue = builder.CreateZExt(oldValue, varty);
+
+    return PushReturn(oldValue);
+  });
+}
+
+IN_ERROR innative::Compiler::CompileAtomicCmpXchg(Instruction& ins, WASM_TYPE_ENCODING varTy, const char* name)
+{
+  return CompileAtomicMemInstruction(ins, varTy, varTy, [&](unsigned align, llvmVal* ptr, llvmVal* cmp, llvmVal* newVal) {
+    auto varsize = unsigned(4 * -varTy); // TE_i32(-1) => 4; TE_i64(-2) => 8;
+    auto varty   = builder.getIntNTy(varsize * 8);
+    auto memty   = builder.getIntNTy(align * 8);
+
+    if(varsize > align)
+    {
+      cmp    = builder.CreateTrunc(cmp, memty);
+      newVal = builder.CreateTrunc(newVal, memty);
+    }
+
+    auto cmpxchg = builder.CreateAtomicCmpXchg(ptr, cmp, newVal, SeqCst, SeqCst);
+    cmpxchg->setName(name);
+
+    llvmVal* loaded = builder.CreateExtractValue(cmpxchg, 0);
+    if(varsize > align)
+      loaded = builder.CreateZExt(loaded, varty);
+
+    return PushReturn(loaded);
+  });
+}
+
+using namespace innative::atomic_details;
+
+IN_ERROR innative::Compiler::CompileAtomicInstruction(Instruction& ins)
+{
+  using RmwOp = llvm::AtomicRMWInst::BinOp;
```

Comment:
I assume this is scoped to just this function, is that correct?

Reply:
Yeah, this is just scoped to the function to cut down on the noise for accessing that enum